### PR TITLE
refactor: tighten callRpc response typing

### LINKED ISSUE #159: Tighten callRpc response typing and remove downstream callRpc<any> usage
## Summary
The request-side quick win for `src/lib/rpc-client.ts` is now in place: `callRpc` only accepts object-shaped `args`, and local RPC error parsing no longer relies on `as any`.

The next high-leverage follow-up is on the response side.

## Problem
- Many callers still use `callRpc<any>` or otherwise rely on weak response typing.
- This keeps `any` flowing through report hooks, maintenance hooks, and other RPC consumers even after the request-side boundary is tightened.

## Scope
- Identify highest-traffic `callRpc<any>` sites
- Replace them with concrete DTOs or `unknown` + narrowing helpers where practical
- Prioritize runtime paths before test scaffolding
- Keep request-side behavior unchanged

## Suggested starting points
- `src/hooks/use-cached-maintenance.ts`
- `src/app/(app)/reports/hooks/use-inventory-data.ts`
- `src/hooks/use-cached-equipment.ts`

## Verification
- `node scripts/npm-run.js run verify:no-explicit-any`
- `node scripts/npm-run.js run typecheck`
- focused tests for touched hooks/components
- `node scripts/npm-run.js npx react-doctor@latest . --verbose -y --project nextn --offline --diff main`

### PR DESCRIPTION
**Pull request by thienchi2109**
## Summary
- remove the remaining response-side `callRpc<any>` leaf usage in repair-request equipment flows, dashboard equipment attention queries, QR lookup, and edit-equipment updates
- replace downstream `callRpc<any>` usage with concrete contracts or small typed helpers while keeping the RPC wire contract unchanged
- add focused TDD coverage for each wave and remove adjacent explicit `any` from the touched runtime files

## Verification
- `node scripts/npm-run.js run verify:no-explicit-any`
- `node scripts/npm-run.js run typecheck`
- `node scripts/npm-run.js run test:run -- "src/app/(app)/equipment/__tests__/useEquipmentData.test.ts" "src/app/(app)/equipment/__tests__/equipmentMutations.test.ts" "src/hooks/__tests__/use-cached-equipment.role-normalization.test.ts" "src/app/(app)/reports/hooks/__tests__/use-inventory-data.types.test.ts" "src/app/(app)/reports/__tests__/ReportsPage.auth-and-hooks.test.tsx" "src/hooks/__tests__/use-cached-maintenance-rpc.test.ts" "src/hooks/__tests__/use-cached-maintenance-barrel.test.ts" "src/components/__tests__/add-equipment-dialog.queries.test.ts" "src/app/(app)/repair-requests/__tests__/repair-requests-equipment-rpc.test.ts" "src/app/(app)/repair-requests/__tests__/useRepairRequestsDeepLink.test.ts" "src/app/(app)/repair-requests/__tests__/RepairRequestsCreateSheet.test.tsx" "src/components/__tests__/qr-action-sheet.test.tsx" "src/hooks/__tests__/use-dashboard-stats.types.test.ts" "src/components/__tests__/edit-equipment-dialog.rpc.test.ts" "src/components/__tests__/equipment-dialogs.crud.test.tsx"`
- `node scripts/npm-run.js npx react-doctor@latest . --verbose -y --project nextn --offline --diff main`

## Notes
- React Doctor diff score: `98/100`
- Non-blocking warnings remain on pre-existing accessibility/file-size issues in `RepairRequestsCreateSheet` and `QRActionSheet`

Closes #159

<h3>Greptile Summary</h3>

This PR completes the `callRpc<any>` response typing refactor (issue #159), replacing `any`-typed RPC responses with concrete generics or `unknown` + explicit narrowing across equipment caching, inventory reports, maintenance, dashboard stats, QR lookup, repair-request, and dialog-level RPC call sites.

Key improvements:
- **Equipment hooks**: `callRpc<boolean>` for `equipment_update`; `useUpdateEquipment` now correctly invalidates the detail cache instead of writing an unknown boolean into it
- **Inventory mapper extraction** (`use-inventory-data.types.ts`): well-typed narrowers for all five inventory payload shapes; incidentally fixes a double-count bug where completed `thanh_ly` items with both `ngay_ban_giao` and `ngay_hoan_thanh` in range could appear in both the transfer and liquidation sets
- **Maintenance file split**: oversized `use-cached-maintenance.ts` reduced by extracting `use-cached-maintenance.types.ts` and `use-cached-maintenance.rpc.ts`
- **Shared `getUnknownErrorMessage` utility** (`src/lib/error-utils.ts`): consolidates error message extraction across `any` / plain-object / string rejection shapes and is adopted consistently across most touched call sites
- **Type-level assertion files** (`.types.assert.ts`): compile-time safety nets for the new type contracts, consistent with the repo pattern

One minor inconsistency was found: the `fetchInitialData` catch block in `useRepairRequestsDeepLink.ts` uses an inline `error instanceof Error ? error.message : ''` guard instead of `getUnknownErrorMessage`, causing empty toast details for non-Error rejections (plain-object errors, raw string throws). All other catch blocks touched by this PR use the shared utility.

<h3>Confidence Score: 5/5</h3>

Safe to merge; no breaking changes, wire contracts unchanged, test coverage is thorough, and prior review concerns are resolved.

All previous P0/P1 concerns from prior review rounds have been addressed (equipment_update boolean contract, shared error utility, pending liquidation regression). The only remaining finding is a single P2 inconsistency in useRepairRequestsDeepLink where getUnknownErrorMessage is not used, causing empty error details for non-Error rejections in one catch block — this does not affect the happy path and is a follow-up improvement rather than a blocker.

src/app/(app)/repair-requests/_hooks/useRepairRequestsDeepLink.ts — the fetchInitialData catch block should use getUnknownErrorMessage for consistency with the rest of the PR.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/app/(app)/repair-requests/_hooks/useRepairRequestsDeepLink.ts | Migrated from direct callRpc to typed RPC helpers; one catch block uses inline instanceof check instead of the shared getUnknownErrorMessage utility introduced elsewhere in this PR. |
| src/app/(app)/reports/hooks/use-inventory-data.types.ts | New mapper file with well-typed narrowers for all inventory RPC payloads; correctly fixes a double-count bug for completed thanh_ly items that existed in the old code. |
| src/app/(app)/repair-requests/repair-requests-equipment-rpc.ts | Clean new typed RPC helper; eliminates duplicated inline callRpc calls and any-cast mapping from two callers. |
| src/hooks/use-cached-equipment.ts | All callRpc calls properly typed; useUpdateEquipment correctly invalidates the detail cache instead of writing an unknown boolean payload into it. |
| src/components/add-equipment-dialog.queries.ts | Added isTenantListItem guard and numeric coercion for tenant IDs; correctly handles string IDs and filters null/empty names. |
| src/lib/error-utils.ts | New shared utility correctly handles all three error shapes (Error instances, plain objects with message property, raw strings) and provides a consistent fallback. |
| src/components/edit-equipment-dialog.rpc.ts | Clean extraction; callRpc typed as boolean and return value is intentionally discarded (void), consistent with the cache-invalidation approach in useUpdateEquipment. |
| src/hooks/use-dashboard-stats.ts | Types correctly extracted to use-dashboard-stats.types.ts and re-exported to preserve the module surface. |

</details>

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/app/(app)/reports/hooks/use-inventory-data.types.ts`, line 1138-1196 ([link](https://github.com/thienchi2109/qltbyt-nam-phong/blob/4b1ce4d5aa903f32e6b4deb0a7ea6414c164a487/src/app/(app)/reports/hooks/use-inventory-data.types.ts#L1138-L1196)) 

   <a href="#">P2</a> **Intentional behavior change for `thanh_ly` transfers worth noting**

   The new routing in `mapExportedInventoryItems` changes how liquidation-type transfer rows are handled compared to the old code:

   - **Old code**: `thanh_ly` rows were included in *both* `exportedFromTransfers` (when `ngay_ban_giao` was in-range, regardless of `trang_thai`) *and* `exportedFromLiquidation` (when `trang_thai === 'hoan_thanh'` and `ngay_hoan_thanh` was in-range). This could double-count completed liquidations, and would misclassify in-progress ones as `transfer_external`.
   - **New code**: `thanh_ly` rows are routed exclusively to the liquidation branch and only appear when `trang_thai === 'hoan_thanh'` and `ngay_hoan_thanh` is in-range.

   This is almost certainly the correct behavior, but it's a subtle runtime change not covered by the new unit tests (the test only exercises the happy-path of a completed liquidation). Consider adding a test case for a `thanh_ly` row that has `ngay_ban_giao` in range but `trang_thai !== 'hoan_thanh'` to confirm it is intentionally excluded.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/app/(app)/reports/hooks/use-inventory-data.types.ts
   Line: 1138-1196

   Comment:
   **Intentional behavior change for `thanh_ly` transfers worth noting**

   The new routing in `mapExportedInventoryItems` changes how liquidation-type transfer rows are handled compared to the old code:

   - **Old code**: `thanh_ly` rows were included in *both* `exportedFromTransfers` (when `ngay_ban_giao` was in-range, regardless of `trang_thai`) *and* `exportedFromLiquidation` (when `trang_thai === 'hoan_thanh'` and `ngay_hoan_thanh` was in-range). This could double-count completed liquidations, and would misclassify in-progress ones as `transfer_external`.
   - **New code**: `thanh_ly` rows are routed exclusively to the liquidation branch and only appear when `trang_thai === 'hoan_thanh'` and `ngay_hoan_thanh` is in-range.

   This is almost certainly the correct behavior, but it's a subtle runtime change not covered by the new unit tests (the test only exercises the happy-path of a completed liquidation). Consider adding a test case for a `thanh_ly` row that has `ngay_ban_giao` in range but `trang_thai !== 'hoan_thanh'` to confirm it is intentionally excluded.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/app/(app)/repair-requests/_hooks/useRepairRequestsDeepLink.ts
Line: 825

Comment:
**Inline error extraction misses non-`Error` rejections**

This catch block uses `error instanceof Error ? error.message : ''` but the shared `getUnknownErrorMessage` utility was extracted specifically for this PR to handle all three cases: `Error` instances, plain objects with a `message` property, and raw strings. As written, a rejection like `{ message: 'Permission denied' }` or a bare `'Network error'` string would produce an empty suffix in the toast description, while the same rejection in `useUpdateEquipment`, `QRActionSheet`, and `AddMaintenancePlanDialog` (all touched by this PR) would display the full message.

```suggestion
          description: 'Không thể tải danh sách thiết bị. ' + getUnknownErrorMessage(error),
```

You'll also need to add the import at the top of the file:
```ts
import { getUnknownErrorMessage } from '@/lib/error-utils'
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (10): Last reviewed commit: ["fix: tighten equipment dialog and role g..."](https://github.com/thienchi2109/qltbyt-nam-phong/commit/b2f1bf6dbd339914d7615f16612dee7c59532d9b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26653770)</sub>

### CODE REVIEW

File: src/app/(app)/reports/hooks/use-inventory-data.types.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+      }
+
+      items.push({
+        ...baseItem,
+        ngay_nhap: completedAt,
+        source: 'liquidation',
+        destination: 'Thanh lý',
+      })
+      continue
+    }
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply src/app/(app)/reports/hooks/use-inventory-data.types.ts]: Fixed in f3881b0. I checked the pre-refactor behavior in `origin/main` and you were right: pending `thanh_ly` rows with `ngay_ban_giao` in range were previously reported through the transfer path. `mapExportedInventoryItems` now only `continue`s after it successfully emits a completed liquidation item, and otherwise falls through to the handover-date branch. That restores the old pending-liquidation behavior while still avoiding the old double-count for completed liquidations. I added a regression test in `src/app/(app)/reports/hooks/__tests__/use-inventory-data.types.test.ts` for the pending-liquidation edge case.

File: src/components/edit-equipment-dialog.tsx [hunk 2]
```diff
@@ -136,8 +148,12 @@ export function EditEquipmentDialog({ open, onOpenChange, onSuccess, equipment }
       onSuccess()
       onOpenChange(false)
     },
-    onError: (error: any) => {
-      toast({ variant: 'destructive', title: 'Lỗi', description: 'Không thể cập nhật thiết bị. ' + (error?.message || '') })
+    onError: (error: unknown) => {
+      toast({
+        variant: 'destructive',
+        title: 'Lỗi',
+        description: 'Không thể cập nhật thiết bị. ' + (error instanceof Error ? error.message : ''),
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply src/components/edit-equipment-dialog.tsx]: Fixed in b2f1bf6. `EditEquipmentDialog` now uses `getUnknownErrorMessage(error)` before composing the toast description, so plain-object rejections no longer collapse to `Không thể cập nhật thiết bị. `. I added a regression case in `src/components/__tests__/equipment-dialogs.crud.test.tsx` that exercises a `{ message: "Permission denied" }` rejection, and the submit handler now swallows the rejected `mutateAsync` promise after `onError` runs so the error path no longer leaks an unhandled rejection during submit.

File: src/hooks/use-cached-equipment.ts [hunk 3]
```diff
@@ -2,11 +2,37 @@ import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
...
 export const equipmentKeys = CacheKeys.equipment
 
+type EquipmentPayload = Record<string, unknown>
+type CacheScopeUser = Pick<User, 'role' | 'khoa_phong'>
+
+function normalizeSessionRole(role: string | null | undefined): UserRole | null {
+  if (!role) return null
+
+  const normalized = role.trim().toLowerCase()
+  return normalized in USER_ROLES ? (normalized as UserRole) : null
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply src/hooks/use-cached-equipment.ts]: Fixed in b2f1bf6. `normalizeSessionRole` now uses `Object.hasOwn(USER_ROLES, normalized)` instead of the prototype-walking `in` operator, so strings like `constructor` are no longer cast as valid `UserRole`s. I extended `src/hooks/__tests__/use-cached-equipment.role-normalization.test.ts` with a regression case that verifies prototype-property names are rejected and `DepartmentCacheUtils.getUserCacheScope` receives `null` for that session.
